### ISSUE TITLE
What are the intended bounds semantics for vectorized get_at coordinates?

### ISSUE THREAD

**Issue by Gravifer** [fferflo/einx#37]
Hi! I ran into an indexing behavior that made me wonder what coordinate domain `get_at` intends to support.

Using PyTorch:

```python
import einx
import torch

x = torch.arange(10).reshape(2, 5)
# tensor([[0, 1, 2, 3, 4],
#         [5, 6, 7, 8, 9]])

print(
    einx.get_at(
        "batch [classes], batch -> batch",
        x,
        torch.tensor([5, 0]),
    )
)
```

With `einx==0.4.2`, this produces:

```text
tensor([5, 5])
```

The first coordinate, `5`, is out of bounds for the local `classes` axis of length 5. A natural reading of the expression would be the elementary operation:

```python
output[b] = x[b, target[b]]
```

Under that interpretation, the first lookup would be out of bounds rather than selecting `x[1, 0]`.

Negative coordinates expose a related ambiguity:

```python
target = torch.tensor([-1, -1])

einx.get_at("batch [classes], batch -> batch", x, target)
# tensor([9, 4])
```

Applying PyTorch indexing separately to each batch entry would instead produce:

```python
torch.stack([x[b, target[b]] for b in range(2)])
# tensor([4, 9])
```

I believe the current result follows from the ravelled lowering. For a tensor of shape `(2, 5)`, the flat coordinate is calculated as:

```text
5 * batch + target
```

Consequently:

```text
target = [ 5, 0]  -> flat coordinates [ 5, 5]
target = [-1,-1]  -> flat coordinates [-1, 4]
```

Those coordinates are then applied to the globally flattened tensor. An axis-local out-of-range coordinate can therefore carry or borrow into another vectorized slice while remaining a valid flat coordinate.

For nonnegative in-bounds coordinates, the ravelled formulation is equivalent to applying the elementary indexing operation separately. The current tests appear to exercise that domain.

Reading the pre-0.4 implementation, it appears that indexing was performed as an elementary operation under `vmap`. That would have preserved the boundary of each indexed slice and inherited the backend's indexing behavior separately for each invocation.

I do not necessarily think that reverse indexing needs to be supported; defining the supported domain as nonnegative, per-axis in-bounds coordinates—and treating reverse or out-of-range indexing as unsupported—would also resolve the ambiguity cleanly.

> *Environment*
> - einx 0.4.2
> - torch 2.11.0

**Comment by fferflo** [comment]
Hi! Thanks for bringing this up.

The semantics in get_at (and also set_at, add_at, etc) currently is that the behavior for out-of-bounds coordinates is undefined (i.e. the user promises that all coordinates are in bounds). This is, however, not documented properly - I'll add a comment to the function documentation to clarify this.

A good extension would probably be to add a `mode` argument to these indexing operations (maybe similar to the mode argument [here](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.ndarray.at.html)). I'll see if this is possible for the next version release of einx.
